app/services/dark-mode.ts

```diff
@@ -0,0 +1,135 @@
+    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
+
+    // Create a listener for media query result changes
+    const mqChangeListener = (e: MediaQueryListEvent) => {
+      this.systemPreference = e.matches ? 'dark' : 'light';
+      this.invokeUpdateListeners();
+    };
+
+    // Load the current system preference by executing the media query
+    this.systemPreference = mediaQuery.matches ? 'dark' : 'light';
+
+    // Subscribe to media query result changes
+    mediaQuery.addEventListener('change', mqChangeListener);
+
+    // Unsubscribe from media query changes when service is destroyed
+    registerDestructor(this, () => {
+      mediaQuery.removeEventListener('change', mqChangeListener);
+    });
+  }
+
+  /**
+   * Indicates whether Dark Mode is currently enabled: either because visiting a
+   * "dark route", or derived from localStorage & system preference
+   */
+  get isEnabled(): boolean {
+    return this.#isVisitingDarkRoute || this.#isLocalStorageEnabled;
+  }
+
+  /**
+   * Returns whether Dark Mode should be enabled based on localStorage & system preference
+   * @private
+   */
+  get #isLocalStorageEnabled(): boolean {
+    return this.localStoragePreference === 'dark' || (this.localStoragePreference === 'system' && this.systemPreference === 'dark');
+  }
+
+  /**
+   * Returns whether Dark Mode should be enabled because the user is visiting a "dark route"
+   * @private
+   */
+  get #isVisitingDarkRoute(): boolean {
+    return this.router.currentRoute.metadata instanceof RouteInfoMetadata && !!this.router.currentRoute.metadata.isDarkRoute;
+  }
+
+  /**
+   * Invokes all currently registered update listeners, passing them a new value
+   */
+  @action invokeUpdateListeners() {
+    for (const listener of this.#registeredListeners) {
+      listener(this.isEnabled);
+    }
+  }
+
+  /**
+   * Registers a callback to invoke when Dark Mode environment changes
+   * @param listener method to invoke, new boolean value will be passed as first argument
+   */
+  @action registerUpdateListener(listener: UpdateListener) {
+    this.#registeredListeners.push(listener);
+  }
+
+  /**
+   * Unregisters a callback previsouly resigtered with `registerUpdateListener`
+   * @param listener method to unregister, must be exactly the same instance, compared using `===`
+   */
+  @action unregisterUpdateListener(listener: UpdateListener) {
+    if (this.#registeredListeners.includes(listener)) {
+      this.#registeredListeners.removeAt(this.#registeredListeners.indexOf(listener));
+    }
+  }
+
+  /**
+   * Writes the new Dark Mode preference to localStorage and executes all currently
+   * registered update listeners, passing them a new value
+   * @param {'system'|'dark'|'light'|null} newValue New dark mode preference
+   */
+  @action updateLocalStoragePreference(newValue?: LocalStoragePreference) {
+    if (!newValue) {
+      this.localStorage.removeItem(LOCAL_STORAGE_KEY);
```

Comment:
I don't think we've settled on proper local storage isolation in the tests environment. And this will require writing the preference to real local storage during tests — which messes up things for `localhost:4200` (at least)